runtime/src/loader_utils.rs

```diff
@@ -313,6 +256,91 @@ pub fn set_upgrade_authority<T: Client>(
         .unwrap();
 }
 
+pub fn instructions_to_load_program_of_loader_v4<T: Client>(
+    bank_client: &T,
+    payer_keypair: &Keypair,
+    authority_keypair: &Keypair,
+    name: &str,
+    program_keypair: Option<Keypair>,
+    target_program_id: Option<&Pubkey>,
+) -> (Keypair, Vec<Instruction>) {
+    let mut instructions = Vec::new();
+    let loader_id = &loader_v4::id();
+    let program = load_program_from_file(name);
+    let program_keypair = program_keypair.unwrap_or_else(|| {
+        let program_keypair = Keypair::new();
+        instructions.push(system_instruction::create_account(
+            &payer_keypair.pubkey(),
+            &program_keypair.pubkey(),
+            bank_client
+                .get_minimum_balance_for_rent_exemption(program.len())
+                .unwrap(),
+            0,
+            loader_id,
+        ));
+        program_keypair
+    });
+    instructions.push(loader_v4::truncate_uninitialized(
+        &program_keypair.pubkey(),
+        &authority_keypair.pubkey(),
+        program.len() as u32,
+        &payer_keypair.pubkey(),
+    ));
+    let chunk_size = CHUNK_SIZE;
+    let mut offset = 0;
+    for chunk in program.chunks(chunk_size) {
+        instructions.push(loader_v4::write(
+            &program_keypair.pubkey(),
+            &authority_keypair.pubkey(),
+            offset,
+            chunk.to_vec(),
+        ));
+        offset += chunk_size as u32;
+    }
+    instructions.push(if let Some(target_program_id) = target_program_id {
+        loader_v4::deploy_from_source(
+            target_program_id,
+            &authority_keypair.pubkey(),
+            &program_keypair.pubkey(),
+        )
+    } else {
+        loader_v4::deploy(&program_keypair.pubkey(), &authority_keypair.pubkey())
+    });
+    (program_keypair, instructions)
+}
+
+pub fn load_program_of_loader_v4(
+    bank_client: &mut BankClient,
+    bank_forks: &RwLock<BankForks>,
+    payer_keypair: &Keypair,
+    authority_keypair: &Keypair,
+    name: &str,
+) -> (Arc<Bank>, Pubkey) {
+    let (program_keypair, instructions) = instructions_to_load_program_of_loader_v4(
+        bank_client,
+        payer_keypair,
+        authority_keypair,
+        name,
+        None,
+        None,
+    );
+    let signers: &[&[&Keypair]] = &[
+        &[payer_keypair, &program_keypair],
+        &[payer_keypair, &program_keypair, authority_keypair],
+        &[payer_keypair, authority_keypair],
+    ];
+    for (index, instruction) in instructions.into_iter().enumerate() {
+        let message = Message::new(&[instruction], Some(&payer_keypair.pubkey()));
+        bank_client
+            .send_and_confirm_message(signers[index.min(signers.len() - 1)], message)
```

Comment:
Some documentation will be helpful here, as to why we are doing `signers[index.min(signers.len() - 1)]`. I am able to understand it right now, but may not be obvious for the future maintainers.

Reply:
I reimplemented it using iterators instead of indices, hopefully that makes it self explanatory.